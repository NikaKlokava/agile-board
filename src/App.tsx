import { Route, Routes } from "react-router";
import "./index.css";
import { BoardPage } from "./pages/board";
import { ErrorPage } from "./pages/error";

function App() {
  return (
    <Routes>
      <Route path="/agile-board" element={<BoardPage />} />
      <Route path="*" element={<ErrorPage />} />
    </Routes>
  );
}

export default App;
