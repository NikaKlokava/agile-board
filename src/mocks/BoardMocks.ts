type MockSubtaskType = {
  name: string;
};

type MockTaskType = {
  name: string;
  description: string;
  subtasks: Array<MockSubtaskType>;
};

export const MockTask: MockTaskType = {
  name: "Create something",
  description: "... it's very important ...",
  subtasks: [
    { name: "while doing 1..." },
    { name: "while doing 2..." },
    { name: "while doing 3..." },
  ],
};

export const MockBoards = [
  {
    name: "Platform Launch",
    board_column: [
      { title: "Todo", tasks: Array.from({ length: 4 }, () => MockTask) },
      { title: "Doing", tasks: Array.from({ length: 6 }, () => MockTask) },
      { title: "Done", tasks: Array.from({ length: 2 }, () => MockTask) },
    ],
  },
  {
    name: "Marketing Plan",
    board_column: [
      { title: "Todo", tasks: Array.from({ length: 7 }, () => MockTask) },
      { title: "Doing", tasks: Array.from({ length: 2 }, () => MockTask) },
      { title: "Done", tasks: Array.from({ length: 4 }, () => MockTask) },
    ],
  },
  {
    name: "Roadmap",
    board_column: [
      { title: "Now", tasks: Array.from({ length: 4 }, () => MockTask) },
      { title: "Next", tasks: Array.from({ length: 6 }, () => MockTask) },
    ],
  },
];
