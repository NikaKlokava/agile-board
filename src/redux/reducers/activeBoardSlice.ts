import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";

const initialState: BoardType = {
  uuid: "",
  name: "",
  columns: [],
  usersEmail: [],
  time: 0,
};

export const activeBoardSlice = createSlice({
  name: "active-board",
  initialState,
  reducers: {
    selectBoard: (_, action: PayloadAction<BoardType>) => action.payload,
    deleteActiveBoard: (state: BoardType) => ({ ...state, ...initialState }),
  },
});

export const { selectBoard, deleteActiveBoard } = activeBoardSlice.actions;

export default activeBoardSlice.reducer;
