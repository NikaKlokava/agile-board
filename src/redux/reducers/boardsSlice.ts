import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import { v4 as uuidv4 } from "uuid";

const initialState: BoardsType | never[] = {
  boards: [],
};

export const boardsSlice = createSlice({
  name: "boards",
  initialState,
  reducers: {
    addBoard: (
      state: BoardsType,
      action: PayloadAction<{ name: string; columns: { title: string }[] }>
    ) => {
      const newBoard = {
        ...action.payload,
        uuid: uuidv4(),
        columns: action.payload.columns.map((column) => {
          return {
            ...column,
            uuid: uuidv4(),
          };
        }),
      };
      state.boards = [...state.boards, newBoard];
    },
    addNewColumn: (
      state: BoardsType,
      action: PayloadAction<{
        name: string;
        columns: { title: string; uuid?: string }[];
        uuid: string | undefined;
      }>
    ) => {
      const newColumns = action.payload.columns.map((column) => {
        const newColumn = {
          ...column,
          uuid: uuidv4(),
        };
        if (!column.uuid) return newColumn;
        return column;
      });
      state.boards = state.boards.map((board: BoardType) => {
        const updatedBoard = {
          ...board,
          name: action.payload.name,
          columns: newColumns,
        };
        if (board.uuid === action.payload.uuid) return updatedBoard;
        return board;
      });
    },
    deleteBoard: (
      state: BoardsType,
      action: PayloadAction<{ uuid: string }>
    ) => {
      state.boards = state.boards.filter(
        (board: BoardType) => board.uuid !== action.payload.uuid
      );
    },
  },
});

export const { addBoard, addNewColumn, deleteBoard } = boardsSlice.actions;

export default boardsSlice.reducer;
