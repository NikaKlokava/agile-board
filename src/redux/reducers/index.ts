import { combineReducers } from "redux";
import { activeBoardReducer } from "./activeBoardReducer";
import { boardsReducer } from "./boardsReducer";
import { tasksReducer } from "./tasksReducer";

export const rootReducer = combineReducers({
  boards: boardsReducer,
  tasks: tasksReducer,
  activeBoard: activeBoardReducer,
});
