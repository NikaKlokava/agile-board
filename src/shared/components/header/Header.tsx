import { Button } from "../button";
import cl from "./header.module.css";

export const Header = () => {
  return (
    <div className={cl.header_wrapper}>
      <div className={cl.app_title}>
        <div className={cl.app_logo}></div>
        <h1 className={cl.title}>AGILE-BOARD</h1>
        <p className={cl.board_name}>Board name</p>
        <div className={cl.navbar}></div>
      </div>
      <div className={cl.options}>
        <Button text={"Add New Task"} />
        <div className={cl.options_icon}></div>
      </div>
    </div>
  );
};
