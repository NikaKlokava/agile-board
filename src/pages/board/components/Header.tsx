import { useState } from "react";
import { MockBoards } from "../../../mocks/BoardMocks";
import { Button } from "../../../shared/components/button";
import { EditBoardModal, NewTaskModal } from "./modals";
import { DeleteBoardModal } from "./modals/DeleteBoardModal";
import { OptionsModal } from "./modals/OptionsModal";
import cl from "./styles/header.module.css";

export const Header = () => {
  const [newTaskVisible, setNewTaskVisible] = useState<boolean>(false);
  const [optionsVisible, setOptionsVisible] = useState<boolean>(false);
  const [editBoardVisible, setEditBoardVisible] = useState<boolean>(false);
  const [deleteBoardVisible, setDeleteBoardVisible] = useState<boolean>(false);

  const data = MockBoards;
  return (
    <div className={cl.header_wrapper}>
      <div className={cl.app_title}>
        <div className={cl.app_logo}></div>
        <h1 className={cl.title}>AGILE-BOARD</h1>
        <p className={cl.board_name}>{data[0].name}</p>
        <div className={cl.navbar}></div>
      </div>
      <div className={cl.options}>
        <Button
          text={"Add New Task"}
          withIcon={true}
          onClick={() => setNewTaskVisible(true)}
          testid={"add-new-task-btn"}
        />
        <div
          className={cl.options_icon}
          data-testid="options-icon"
          onClick={() => setOptionsVisible((prev) => !prev)}
        />
      </div>
      {newTaskVisible && <NewTaskModal />}
      {optionsVisible && !editBoardVisible && (
        <OptionsModal
          onEditClick={() => setEditBoardVisible(true)}
          onDeleteClick={() => setDeleteBoardVisible(true)}
        />
      )}
      {editBoardVisible && <EditBoardModal />}
      {deleteBoardVisible && <DeleteBoardModal />}
    </div>
  );
};
