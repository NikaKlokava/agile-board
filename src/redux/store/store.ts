import { legacy_createStore } from "redux";
import { boardsReducer } from "../reducers/boardReducer";

const store = legacy_createStore(boardsReducer);

export default store;

// export type RootState = ReturnType<typeof store.getState>

// export interface RootState {
//     user: UserState;
//     loader: LoaderState;
//   }
// export type AppDispatch = typeof store.dispatch
