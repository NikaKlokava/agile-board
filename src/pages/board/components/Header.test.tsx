import { render, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { Header } from "./Header";
import App from "../../../App";
import { Provider } from "react-redux";
import store from "../../../redux/store/store";
import { addBoard, changeStatus } from "../../../redux/reducers/boardsSlice";
import { MockBoard, MockBoardWithoutColumn } from "../../../mocks/BoardMocks";
import { selectBoard } from "../../../redux/reducers/activeBoardSlice";

jest.mock("../../../shared/hooks/useAuthorization", () => ({
  useAuthorization: () => ({ isUserExist: true }),
}));

beforeEach(() => jest.clearAllMocks());

describe("Test the Header component", () => {
  test("The Header renders correctly", () => {
    const headerSnap = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <Header />
          </Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(headerSnap).toMatchSnapshot();
  });
});
describe("Test the button Add New Task", () => {
  test("The button Add New Task should be visible when you open the page and have at least one column", () => {
    render(<App />);
    act(() => {
      store.dispatch(changeStatus({ isLoading: false }));
      store.dispatch(addBoard(MockBoard));
      store.dispatch(selectBoard(MockBoard));
    });

    const addNewTaskBtn = screen.queryByTestId("add-new-task-btn");

    expect(addNewTaskBtn).toBeInTheDocument();
  });

  test("The button Add New Task should not be visible if have not at least one column", () => {
    render(<App />);
    act(() => {
      store.dispatch(addBoard(MockBoardWithoutColumn));
      store.dispatch(selectBoard(MockBoardWithoutColumn));
    });

    const addNewTaskBtn = screen.queryByTestId("add-new-task-btn");
    expect(addNewTaskBtn).not.toBeInTheDocument();
  });

  test("The NewTaskModal should be visible on button Add New Task click", () => {
    render(<App />);
    act(() => {
      store.dispatch(selectBoard(MockBoard));
    });
    const addNewTaskBtn = screen.getByTestId("add-new-task-btn");

    act(() => {
      addNewTaskBtn.click();
    });

    const newTaskModal = screen.queryByTestId("new-task-modal");

    expect(newTaskModal).toBeInTheDocument();
  });
});

describe("Test the icon Options", () => {
  test("The OptionsModal should be visible by clicking", () => {
    render(<App />);
    const optionsIcon = screen.getByTestId("options-icon");

    act(() => {
      optionsIcon.click();
    });

    const optionsModal = screen.queryByTestId("options-modal");
    expect(optionsModal).toBeInTheDocument();
  });
});
