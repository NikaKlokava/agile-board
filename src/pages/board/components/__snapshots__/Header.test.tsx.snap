// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Header component The Header renders correctly 1`] = `
<div
  className="header_wrapper"
>
  <div
    className="app_title"
  >
    <div
      className="app_logo"
    />
    <h1
      className="title"
    >
      AGILE-BOARD
    </h1>
    <p
      className="board_name"
    >
      Platform Launch
    </p>
    <div
      className="navbar"
    />
  </div>
  <div
    className="options"
  >
    <div
      className="button"
      data-testid="add-new-task-btn"
      onClick={[Function]}
    >
      <div
        className="button_icon"
      />
      <div
        className="button_text"
      >
        Add New Task
      </div>
    </div>
    <div
      className="options_icon"
      data-testid="options-icon"
      onClick={[Function]}
    />
  </div>
</div>
`;
